references by uuid

ElectricMaxxx opened the pull request:
This should wrap up #493 

didn't implement the stuff, commented in code not to change.

But there are still the [Questions](https://github.com/doctrine/phpcr-odm/pull/493#issuecomment-43854355) of @dbu

I go top to button through the text/conversation, cause it is really hard to remember what the purpose was, even for me :-)

ElectricMaxxx committed:
use uuid build reference and register document by uuid

dantleech commented on lib/Doctrine/ODM/PHPCR/DocumentManager.php:
```diff
@@ -917,13 +917,11 @@ public function initializeObject($document)
      */
     public function getNodeForDocument($document)
     {
-        if (!is_object($document)) {
```
Why do we remove this validation?

dantleech commented on lib/Doctrine/ODM/PHPCR/DocumentManager.php:
```diff
@@ -917,13 +917,11 @@ public function initializeObject($document)
     {
-        if (!is_object($document)) {
-            throw new InvalidArgumentException('Parameter $document needs to be an object, '.gettype($document).' given');
+        if (!$identifier = $this->unitOfWork->getDocumentId($document)) {
```
`if (null === $identifier = $this->...`

dantleech commented on lib/Doctrine/ODM/PHPCR/DocumentManager.php:
```diff
@@ -917,13 +917,11 @@ public function initializeObject($document)
-        if (!is_object($document)) {
-            throw new InvalidArgumentException('Parameter $document needs to be an object, '.gettype($document).' given');
+        if (!$identifier = $this->unitOfWork->getDocumentId($document)) {
+            throw new InvalidArgumentException('This document is not managed by this manager.');
```
`sprintf('Document of class "%s" is not managed', get_class($document))`

dantleech commented on lib/Doctrine/ODM/PHPCR/UnitOfWork.php:
```diff
@@ -674,7 +675,7 @@ public function getOrCreateProxy($targetId, $className, $locale = null)
         $proxyDocument = $this->dm->getProxyFactory()->getProxy($className, array($metadata->identifier => $targetId));
 
-        // register the document under its own id
+        // register the document under its own id/uuid
```
`path/uuid`.

dantleech commented on lib/Doctrine/ODM/PHPCR/UnitOfWork.php:
```diff
@@ -3113,6 +3126,20 @@ public function getDocumentById($id)
     }
 
     /**
+     * Creates a node from a given path or an uuid
```
`given path or UUID`

dantleech commented on lib/Doctrine/ODM/PHPCR/UnitOfWork.php:
```diff
@@ -3063,18 +3069,25 @@ private function unregisterDocument($document)
+        } catch (ItemNotFoundException $e) {
+            $frozen = false;
+        } catch (PathNotFoundException $e) {
+            $frozen = false;
```
`$frozen = false` can be moved before the `try` block (doesn't need to be repeated twice)

dantleech commented on tests/Doctrine/Tests/ODM/PHPCR/UnitOfWorkTest.php:
```diff
@@ -74,54 +75,66 @@ public function setUp()
             "jcr:system" => array(),
             'username' => $username,
         );
+        if (null !== $uuid) {
```
I prefer a blank line before most if statements.

dantleech commented on tests/Doctrine/Tests/ODM/PHPCR/UnitOfWorkTest.php:
```diff
@@ -74,54 +75,66 @@ public function setUp()
+                ->will($this->returnValue($node));
+        }
+
+
```
One blank line too many

dantleech reviewed (commented):
Can you summarize the motivation behind this PR in the description? I don't really know the full implications of this change.

ElectricMaxxx committed:
try to introduce uuid -> path mapping with documents managed by uuid

ElectricMaxxx commented:
This will fail for sure now ... Did a lot of changes

* introduced a object hash - uuid map, which is filled when a document is registered by its uuid (on `DocumentManager::getReference()` call with proxies referencing documents by their uuid <- @dantleech purpose)
* introduced a uuid - path map to find documents by their path in identity map, which are managed by uuid

so all documents still live in `identityMap` by their id, but the id can be the absolute paths directly or uuids which needs to be converted into paths (uuid - id map).

ElectricMaxxx committed:
change method name

ElectricMaxxx committed:
fix test driven node existence

dantleech commented on lib/Doctrine/ODM/PHPCR/UnitOfWork.php:
```diff
@@ -100,6 +101,18 @@ class UnitOfWork
     /**
+     * @var array
+     */
+    private $documentUuids = array();
```
Ideally this would be better named `$oidToUuidMap`, but to be consistent this naming is fine for now I guess. But it should be commented.

dantleech commented on lib/Doctrine/ODM/PHPCR/UnitOfWork.php:
```diff
@@ -100,6 +101,18 @@ class UnitOfWork
+     *
+     * @var array
+     */
+    private $uuidIdMapping = array();
```
`Id` in `uuidId` is redundant. Better to name `$uuidToPathMap`.

dantleech commented on lib/Doctrine/ODM/PHPCR/UnitOfWork.php:
```diff
@@ -3058,23 +3076,43 @@ private function unregisterDocument($document)
             $this->documentVersion[$oid]
         );
 
+        if (!empty($this->documentUuids[$oid]) && array_key_exists($this->documentUuids[$oid], $this->uuidIdMapping)) {
```
Why check for `!empty`?

dantleech reviewed (commented):
Still not sure what the *motivation* of this is? Regarding the mapping I wonder if it is worth introducing a Registry class to encapsulate the various maps, reducing complication in this already very complicated class, something *like*: https://github.com/sulu/sulu-document-manager/blob/master/lib/DocumentRegistry.php.

But maybe something for a later iteration.

ElectricMaxxx committed:
add one test to the crud tests
